AvaloniaUI/Avalonia
# MacOS sandboxing feature

## What does the pull request do?

This PR resurrects idea of old PR - https://github.com/AvaloniaUI/Avalonia/pull/6540.
In short, in order to make macOS app working with the AppStore sandbox, developers need to be careful with how they interact with the File System:
- Each operation **should** be wrapped in secure scope session - `[NSURL startAccessingSecurityScopedResource]` + `[NSURL stopAccessingSecurityScopedResource]`.
- When any file or folder are opened via picker, developers need to `bookmark` access to these files in order to reuse them between app sessions. Simply saving full file path won't work, as OS might not allow direct file access.

Old PR was created when we didn't have a better idea how this API should look like. And later it was implemented for mobile and browser platforms with new StorageProvider API, where we fully support sandboxed bookmarks. macOS platform was temporary skipped due to complexity and chance of introducing bugs to the existing backend (+it was low priority).

## How to test

1. Open ControlCatalog, Dialogs page.
2. Open any file or folder
3. Copy "bookmark" text field
4. Reopen the app, dialogs page, input the same bookmark value.
5. Press "read folder/file bookmark".
6. The same folder/file should be opened.
7. Additionally, all other storage related operations should function as they did before without regressions.

Extra: use https://stackoverflow.com/questions/24947661/os-x-app-testing-for-sandbox-violations

## Breaking changes

Two breaking changes:
1. BclStorageFile/Folder.SaveBookmarkAsync now returns encoded string instead of original path (only Windows and Linux). MacOS implementation also returns an actual bookmark from the OS.
9. By default, macOS now uses `[NSURL startAccessingSecurityScopedResource]` each time when user accesses file, including stream operations (we have a stream wrapper for that). It's not a breaking change by itself, but it's rather a big impact that might introduce unwanted bugs. It's possible to disable this behavior by passing `AvaloniaNativePlatformOptions.AppSandboxEnabled = false`, which will revert to old non-sandboxed APIs.

## Fixed issues

Fixes #6537
Reopens #6540
